### DISCUSSION TITLE
BUG: `INSERT ... ON CONFLICT` breaks on magic cookie 'updated value' for column types that cannot convert the string

### DISCUSSION THREAD

**Discussion by h-vetinari** [sqlalchemy/sqlalchemy#10682] [Usage Questions] [answered]
### Describe the bug

The [docs](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert) for `INSERT…ON CONFLICT` provide an example for the most commonly desired upsert behaviour: if the constraint matches, use the incoming values. This is especially important for batch operations.

The example uses `set_=dict(data='updated value')` to update a single column called `data`, and while the docs don't go into much detail about the magic cookie `'updated value'`, it's clear from the generated SQL query that this is not taken as a literal, but rather causes a `SET data = %(param_1)s`.

Trying to adapt this to my example, I tried to set `'updated value'` basically for all columns of my table. Unfortunately, this failed. After some debugging, it appears to be primarily a problem for column types that cannot convert from the string `'updated value'`, i.e. probably there is a layering mismatch between when the types are checked and when the magic cookie gets utilized as "use the incoming values".

In my case, the generated query looked as follows (linebreaks added for legibility):

<details>
<summary>generated query</summary>

```
[SQL:
INSERT INTO
    schema.table
    (col1, col2, col3, col4, col5, col6, col7, col8, col9, col10, col11, col12, col13)
VALUES
    (%(col1_m0)s::VARCHAR, %(col2_m0)s::VARCHAR, %(col3_m0)s::VARCHAR, %(col4_m0)s::VARCHAR,
     %(col5_m0)s::TIMESTAMP WITHOUT TIME ZONE, %(col6_m0)s::VARCHAR, %(col7_m0)s,
     %(col8_m0)s::VARCHAR, %(col9_m0)s::VARCHAR, %(col10_m0)s::VARCHAR, %(col11_m0)s::VARCHAR,
     %(col12_m0)s::JSON, %(col13_m0)s::JSONB),
    (%(col1_m1)s::VARCHAR, %(col2_m1)s::VARCHAR, %(col3_m1)s::VARCHAR, %(col4_m1)s::VARCHAR,
     %(col5_m1)s::TIMESTAMP WITHOUT TIME ZONE, %(col6_m1)s::VARCHAR, %(col7_m1)s,
     %(col8_m1)s::VARCHAR, %(col9_m1)s::VARCHAR, %(col10_m1)s::VARCHAR, %(col11_m1)s::VARCHAR,
     %(col12_m1)s::JSON, %(col13_m1)s::JSONB),
    [...]
ON CONFLICT ON CONSTRAINT table_primary_key
DO UPDATE SET
    col1 = %(param_1)s::VARCHAR,
    col2 = %(param_2)s::VARCHAR,
    col3 = %(param_3)s::VARCHAR,
    col4 = %(param_4)s::VARCHAR,
    col5 = %(param_5)s::TIMESTAMP WITHOUT TIME ZONE,
    col6 = %(param_6)s::VARCHAR,
    col7 = %(param_7)s,
    col8 = %(param_8)s::VARCHAR,
    col9 = %(param_9)s::VARCHAR,
    col10 = %(param_10)s::VARCHAR,
    col11 = %(param_11)s::VARCHAR,
    col12 = %(param_12)s::JSON,
    col13 = %(param_13)s::JSONB
]
```

</details>

Which cause an error for the Boolean `col7`:
```
StatementError: (builtins.TypeError) Not a boolean value: 'updated value'
```
and (after removing `col7`) also for the timestamp column `col5`:
```
DataError: (psycopg.errors.InvalidDatetimeFormat) invalid input syntax for type timestamp: "updated value"
```

Confirming this line of thought, the query succeeds if those two columns are removed from the `set_=dict(...)`!

Given that the literal `'updated value'` is never intended to be taken at face value, but just supposed to be an indicator to use the incoming values, I think it's obvious that this should work for any column type. It probably should also be documented more clearly. Alternatively (or additionally), it might be useful to provide a switch that just allows using the incoming values as-is, without having to construct the corresponding dictionary of magic cookies to pass to `set_=`.

### SQLAlchemy Version in Use

2.0.23

### DBAPI (i.e. the database driver)

psycopg 3

### Database Vendor and Major Version

PostgreSQL 15

### Python Version

3.11

### Operating system

Linux

### To Reproduce

```python
import asyncio
import sqlalchemy as sa
import sqlalchemy.dialects.postgresql

target_table = # from ORM

async def func():
    async with async_session_factory() as session:
        rows = # list_of_dicts
        # get column names from an element (all are dicts with the same keys)
        col_names = [k for k in rows[0].keys()]                                 # fails
        col_names = [k for k in rows[0].keys() if k not in ["col5", "col7"]]    # works

        stmt = sa.dialects.postgresql.insert(target_table).values(rows)
        stmt = stmt.on_conflict_do_update(
            # using the primary key should work by default, but we need to specify constraint
            constraint=target_table.__table__.primary_key,
            # instructions on conflict: use the incoming values for each column
            set_=dict(zip(col_names, ["updated value"] * len(col_names))),
        )
        await session.execute(stmt)
        await session.commit()

asyncio.run(func)
```

### Error

Either of
```
StatementError: (builtins.TypeError) Not a boolean value: 'updated value'

DataError: (psycopg.errors.InvalidDatetimeFormat) invalid input syntax for type timestamp: "updated value"
```

**Comment by CaselIT** [comment]
Hi,

> magic cookie `'updated value'`

there is no magic cookie. `'updated value'` is just an example, set the value to anything you want to update

**Reply by h-vetinari** [reply]
> set the value to anything you want to update

That's not possible for batch updates, where each row has different content, and I want the newest values to be taken on every (conflicting) row, not set a constant value (per column) across all rows.

**Comment by h-vetinari** [comment, marked as answer]
> I want the newest values to be taken on every (conflicting) row, not set a constant value (per column) across all rows.

Despite the oxymoronic name (that's not the fault of sqlalchemy, but postgres), the multiple values that should be updated in the bulk case are found in the `excluded` table.

I.e.
```
set_={c: stmt.excluded.get(c) for c in col_name}
```

does the job that I wanted the magic cookies to do... 🙈

**Reply by CaselIT** [reply]
it's documented here https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#updating-using-the-excluded-insert-values

**Reply by h-vetinari** [reply]
Yes, and I overlooked it because I wasn't interested in "excluded" values. I think the naming is... highly suboptimal, but whatever. Thanks for the help, and sorry for the noise!

**Reply by CaselIT** [reply]
No problem. I think sqlalchemy call it excluded since postgresql does it as well. I agree it's not that meaningful that name
